File: src/core/SDRDictionary.py
from .SDR import *
from .SDRMap import *

class SDRDictionary:

    def __init__(self):
        self.map = SDRMap()
        self.dict = {}
        self.mirrorDict = {}
        self.secondKeys = []

    def set(self,key,value):
        secondKey = SDR(range=2048).indices
        self.secondKeys.append(secondKey)
        self.map.set(key,secondKey)
        self.dict[','.join(map(str,secondKey))] = value
        self.mirrorDict[value] = key
        return key

    def get(self,key):

        if(type(key) is str):
            return self.mirrorDict[key]

        secondKey = self.map.get(key)
        if(len(secondKey) > 8):
            values = []
            for k in self.secondKeys:
                overlap = SDR.Overlap(k,secondKey)
                if(len(overlap) >= 8):
                    values.append(self.dict.get(','.join(map(str,overlap))))
            return values
        return self.dict.get(','.join(map(str,secondKey)))

File: src/core/SDRDictionary.js

const SDR = require('./SDR')
const SDRMap = require('./SDRMap')

module.exports = class SDRDictionary {

    constructor(){
        this.map = new SDRMap()
        this.dict = {}
        this.mirrorDict = {}
        this.secondKeys = []
    }

    set(key,val){
        const secondKey = new SDR({length:2048}).indices
        this.secondKeys.push(secondKey)
        this.map.set(key,secondKey)
        this.dict[secondKey.join()] = val
        this.mirrorDict[val] = key
        return key
    }

    get(key){
        if(typeof key == 'string')
            return this.mirrorDict[key]
        const secondKey = this.map.get(key)
        if(secondKey && secondKey.length > 8){
            const vals = []
            for(var i = 0; i < this.secondKeys.length; i++){
                const overlap = SDR.AND([secondKey,this.secondKeys[i]])
                if(overlap.length >= 8)
                    vals.push(this.dict[overlap.join()])
            }
            return vals
        }
        return this.dict[(secondKey || []).join()] || null
    }

}